File: apps/go/design-pattern/adapter/adapter_test.go
package adapter

import (
	"math"
	"strings"
	"testing"
)

func TestSimpleTextObject(t *testing.T) {
	obj := &SimpleTextObject{
		TextContent: "Hello",
		Size:        1.5,
		ColorValue:  "red",
	}

	if obj.Text() != "Hello" {
		t.Errorf("期待値 'Hello', 実際 %q", obj.Text())
	}
	if obj.SizeInches() != 1.5 {
		t.Errorf("期待値 1.5, 実際 %f", obj.SizeInches())
	}
	if obj.Color() != "red" {
		t.Errorf("期待値 'red', 実際 %q", obj.Color())
	}
}

func TestBritishTextObjectAdapter(t *testing.T) {
	british := &BritishTextObject{
		BritishText:   "Hello",
		SizeMillimeter: 25.4,
		Colour:        "red",
	}
	adapted := &BritishTextObjectAdapter{Object: british}

	if adapted.Text() != "Hello" {
		t.Errorf("期待値 'Hello', 実際 %q", adapted.Text())
	}
	if adapted.Color() != "red" {
		t.Errorf("期待値 'red', 実際 %q", adapted.Color())
	}
}

func TestMillimeterToInchConversion(t *testing.T) {
	british := &BritishTextObject{
		BritishText:   "Test",
		SizeMillimeter: 50.8,
		Colour:        "blue",
	}
	adapted := &BritishTextObjectAdapter{Object: british}

	expected := 2.0
	if math.Abs(adapted.SizeInches()-expected) > 0.001 {
		t.Errorf("期待値 %f, 実際 %f", expected, adapted.SizeInches())
	}
}

func TestRendererWithSimpleObject(t *testing.T) {
	renderer := &Renderer{}
	obj := &SimpleTextObject{
		TextContent: "Hello",
		Size:        1.0,
		ColorValue:  "blue",
	}
	output := renderer.Render(obj)

	if !strings.Contains(output, "Hello") {
		t.Error("出力にテキストが含まれていない")
	}
	if !strings.Contains(output, "blue") {
		t.Error("出力に色が含まれていない")
	}
}

func TestRendererWithAdapter(t *testing.T) {
	renderer := &Renderer{}
	british := &BritishTextObject{
		BritishText:   "Cheerio",
		SizeMillimeter: 25.4,
		Colour:        "crimson",
	}
	adapted := &BritishTextObjectAdapter{Object: british}
	output := renderer.Render(adapted)

	if !strings.Contains(output, "Cheerio") {
		t.Error("出力にテキストが含まれていない")
	}
	if !strings.Contains(output, "crimson") {
		t.Error("出力に色が含まれていない")
	}
	if !strings.Contains(output, "1.00in") {
		t.Errorf("出力にサイズが含まれていない: %s", output)
	}
}

func TestAdapterImplementsInterface(t *testing.T) {
	british := &BritishTextObject{
		BritishText:   "Test",
		SizeMillimeter: 25.4,
		Colour:        "green",
	}

	// Compile-time check that adapter satisfies TextObject interface
	var _ TextObject = &BritishTextObjectAdapter{Object: british}
}

File: apps/typescript/design-pattern/tests/adapter.test.ts
import {
  BritishTextObject,
  BritishTextObjectAdapter,
  Renderer,
} from '../src/adapter';

describe('Adapter パターン', () => {
  it('BritishTextObjectAdapter は text を返す', () => {
    const british = new BritishTextObject('Hello', 25.4, 'red');
    const adapter = new BritishTextObjectAdapter(british);
    expect(adapter.text).toBe('Hello');
  });

  it('BritishTextObjectAdapter は mm を inches に変換する', () => {
    const british = new BritishTextObject('Hello', 25.4, 'red');
    const adapter = new BritishTextObjectAdapter(british);
    expect(adapter.sizeInches).toBeCloseTo(1.0, 5);
  });

  it('BritishTextObjectAdapter は colour を color に変換する', () => {
    const british = new BritishTextObject('Hello', 25.4, 'red');
    const adapter = new BritishTextObjectAdapter(british);
    expect(adapter.color).toBe('red');
  });

  it('Renderer は TextObject を受け取って描画する', () => {
    const british = new BritishTextObject('World', 50.8, 'blue');
    const adapter = new BritishTextObjectAdapter(british);
    const renderer = new Renderer();
    const result = renderer.render(adapter);

    expect(result).toContain('World');
    expect(result).toContain('2.00');
    expect(result).toContain('blue');
  });

  it('BritishTextObject は元のプロパティを保持する', () => {
    const british = new BritishTextObject('Test', 100, 'green');
    expect(british.string).toBe('Test');
    expect(british.sizeMm).toBe(100);
    expect(british.colour).toBe('green');
  });
});
